File: algorithm/sort/insertion/insertion.go
// go run insertion.go

package main

import "fmt"

func main()  {
	arr := insertionSort([]int{1, 4, 5, 6, 3})
	fmt.Printf("result %v", arr)
}

func insertionSort(arr []int) []int {
	for i := range arr {
		perIndex := i - 1
		current := arr[i]

		for perIndex >= 0 && arr[perIndex] > current {
			arr[perIndex + 1] = arr[perIndex]
			perIndex --
		}
		arr[perIndex + 1] = current
	}
	return arr
}

File: algorithm/sort/insertion/insertion.js
// node insertion.js
/**
 * insertion sort
 * @param {Array} arr
 * @return {Array}
 */
function insertionSort(arr) {
    var len = arr.length;

    var preIndex, current;
    for (var i = 0; i < len; i++) {
        current = arr[i];
        preIndex = i - 1;

        while (preIndex >= 0 && arr[preIndex] > current) {
            arr[preIndex + 1] = arr[preIndex];
            preIndex--;
        }
        arr[preIndex + 1] = current;
    }

    return arr;
}

console.log(insertionSort([1, 3, 4, 2, 7]));